File: csharp/1039.minimum-score-triangulation-of-polygon.cs
/*
 * @lc app=leetcode id=1039 lang=csharp
 *
 * [1039] Minimum Score Triangulation of Polygon
 */

using System;

// @lc code=start
public class Solution {

    private static int MinScoreTriangulationRecurse(int[] A, int[,] dp, int l, int r) {
        if (l >= r - 1) {
            return 0;
        }
        if (dp[l, r] > 0) {
            return dp[l, r];
        }
        int ret = int.MaxValue, b = A[l] * A[r];
        for (int m = l + 1; m < r; ++m) {
            ret = Math.Min(ret, b * A[m] + MinScoreTriangulationRecurse(A, dp, l, m) + MinScoreTriangulationRecurse(A, dp, m, r));
        }
        dp[l, r] = ret;
        return ret;
    }

    public int MinScoreTriangulation(int[] A) {
        int n = A.Length;
        return MinScoreTriangulationRecurse(A, new int[n, n], 0, n - 1);
    }
}
// @lc code=end

File: java/1039.minimum-score-triangulation-of-polygon.java
/*
 * @lc app=leetcode id=1039 lang=java
 *
 * [1039] Minimum Score Triangulation of Polygon
 */

// @lc code=start
class Solution {
    public int minScoreTriangulation(int[] values) {
        int n = values.length;
        return helper(values, 0, n - 1, new int[n][n]);
    }

    private static int helper(int[] values, int l, int r, int[][] memo) {
        if (l >= r - 1) {
            return 0;
        }
        if (memo[l][r] > 0) {
            return memo[l][r];
        }
        int base = values[l] * values[r], ret = Integer.MAX_VALUE;
        for (int i = l + 1; i < r; ++i) {
            ret = Math.min(ret, base * values[i] + helper(values, i, r, memo) + helper(values, l, i, memo));
        }
        memo[l][r] = ret;
        return ret;
    }
}
// @lc code=end

